File: app/src/main/java/com/icerockdev/babenko/ui/images/ImagesPresenter.java
package com.icerockdev.babenko.ui.images;

import com.arellomobile.mvp.InjectViewState;
import com.icerockdev.babenko.ui.base.BasePresenter;
import com.icerockdev.babenko.utils.RxUtils;

import timber.log.Timber;

/**
 * Created by Roman Babenko on 10/05/17.
 */
@InjectViewState
public class ImagesPresenter extends BasePresenter<ImagesView> {
    private static final String TAG = "ImagesPresenter";
    private ImagesInteractor mManager;

    public ImagesPresenter(ImagesInteractor manager) {
        mManager = manager;
    }

    @Override
    public void onFirstViewAttach() {
        super.onFirstViewAttach();
        requestPictures();
    }

    private void requestPictures() {
        getViewState().showProgressDialog();
        mManager.requestPicturesList()
                .compose(RxUtils.applyIoMainThreadSchedulersToSingle())
                .doFinally(() -> getViewState().dismissProgressDialog())
                .subscribe(imageItems -> {
                    Timber.tag(TAG).d("Images list length is:%s", imageItems.size());
                    if (imageItems.size() == 0) {
                        getViewState().showListIsEmptyError();
                        return;
                    }
                    getViewState().showImagesList(imageItems);
                }, throwable -> {
                    getViewState().showErrorDialog();
                });
    }

}

File: app/src/main/java/com/icerockdev/babenko/ui/images/ImagesPresenter.kt
package com.icerockdev.babenko.ui.images

import com.arellomobile.mvp.InjectViewState
import com.icerockdev.babenko.applicaiton.utils.applyIoMainThreadSchedulersToSingle
import com.icerockdev.babenko.model.entities.ImageItem
import com.icerockdev.babenko.ui.base.BasePresenter
import timber.log.Timber

/**
 * Created by Roman Babenko on 10/05/17.
 */
@InjectViewState
class ImagesPresenter(private val mManager: ImagesInteractor) : BasePresenter<ImagesView>() {

    public override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        requestPictures()
    }

    private fun requestPictures() {
        viewState.showProgressDialog()
        mManager.requestPicturesList()
            .compose<List<ImageItem>>(applyIoMainThreadSchedulersToSingle<List<ImageItem>>())
            .doFinally { viewState.dismissProgressDialog() }
            .subscribe({ imageItems ->
                Timber.tag(TAG).d("Images list length is:%s", imageItems.size)
                if (imageItems.size == 0) {
                    viewState.showListIsEmptyError()
                } else {
                    viewState.showImagesList(imageItems)
                }
            }) { throwable -> viewState.showErrorDialog() }
    }

    companion object {
        private val TAG = "ImagesPresenter"
    }

}
